### PR TITLE
Andre Marcelo-Tanner - SRE

### PR DESCRIPTION
**Pull request by kzap**
Hi,

Please see the [README](https://github.com/AdaSupport/backend-challenge/blob/65a386dfa0fc8b37f512626151566d94266018d3/README.md) for detailed instructions on how to run this application.

## Running

0. In the `ada-backend-challenge` repo directory, copy `./deployments/docker-compose/.env.sample` to `./.env` and customize as needed.

1. Use `docker-compose up` to run the application.

> If you get an error like:
> `WARNING: The ADA_WEB_HTTP_PORT variable is not set. Defaulting to a blank string.`
>
> This means you need to setup your `.env` file in the same directory. See [README](https://github.com/AdaSupport/backend-challenge/blob/65a386dfa0fc8b37f512626151566d94266018d3/README.md)

```sh
$ docker-compose up
Building web
Step 1/11 : FROM golang:1.12 as build-env
...
Successfully built 76fddb0c8a02
Successfully tagged ada/backend-challenge:andre
```

2. Access the application via your browser: `http://localhost:18080`
Screenshot 2019-04-16 01 48 36

3. Post a message by using the form: `http://localhost:18080/messages`
Screenshot 2019-04-16 01 48 50

4. You will be directed to the conversation page: `http://127.0.0.1:8000/conversations/12345`
Screenshot 2019-04-16 01 48 57

5. You can also POST a message via `curl`:

```sh
curl -H "Content-Type: application/json" -X POST -d '{"conversation_id": "12345", "sender": "Andre" ,"message": "TEST"}' http://localhost:18080/messages

{"status":200,"message":"Success"}
```

6.  You can view the JSON version of the conversation: `http://127.0.0.1:8000/conversations/12345.json`

```json
{"id":"12345","messages":[{"sender":"Andre","message":"Test Message","created":"2019-04-15 17:48:53"}]}
```

## Tests

There is only one test configured due to limit constraints but you run this if you have `Go` installed

```sh
$ cd ./ada-backend-challenge

$ go test -v ./...
?   	github.com/kzap/ada-backend-challenge/cmd/backend-challenge	[no test files]
?   	github.com/kzap/ada-backend-challenge/internal/cmd	[no test files]
?   	github.com/kzap/ada-backend-challenge/internal/config	[no test files]
2019/04/16 02:09:25 /Users/andremarcelo-tanner/Documents/WebPage/AdaSupport/backend-challenge/internal/webserver
=== RUN   TestGetIndex
--- PASS: TestGetIndex (0.00s)
=== RUN   TestGetConversationList
--- PASS: TestGetConversationList (0.00s)
=== RUN   TestGetConversation
--- PASS: TestGetConversation (0.00s)
=== RUN   TestGetConversationJSON
--- PASS: TestGetConversationJSON (0.00s)
=== RUN   TestPostMessages
--- PASS: TestPostMessages (0.00s)
PASS
ok  	github.com/kzap/ada-backend-challenge/internal/webserver	0.009s
```

### CODE REVIEW

File: deployments/docker-compose/.env.sample [hunk 1]
```diff
@@ -0,0 +1,7 @@
+COMPOSE_PROJECT_NAME=ada-backend-challenge
+
+ADA_HOST_IP=10.254.254.254
+
+ADA_WEB_HTTP_PORT=18080
```
Comment (kzap) [inline deployments/docker-compose/.env.sample]: the local port you want to access the web app on

File: deployments/docker-compose/.env.sample [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+ADA_DB_MYSQL_PORT=13306
+ADA_DBADMIN_PORT=18081
```
Comment (kzap) [inline deployments/docker-compose/.env.sample]: ports for using `deployments/docker-compose/docker-compose.dbdev.yml`

File: internal/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"os"
+
+	"github.com/kzap/ada-backend-challenge/internal/config"
+	"github.com/kzap/ada-backend-challenge/internal/webserver"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var rootCmd = &cobra.Command{
```
Comment (kzap) [inline internal/cmd/root.go]: our CLI command with configurable database connection settings

File: internal/webserver/webserver.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+// Route describes the URLs available for this program
+type Route struct {
+	URL         string
+	Description string
+	Methods     string
+	Handler     func(w http.ResponseWriter, r *http.Request)
+}
+
+type Conversation struct {
```
Comment (kzap) [inline internal/webserver/webserver.go]: struct for displaying the conversation data

File: internal/webserver/webserver.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+	ID             int    `json:"-"`
+	ConversationID string `json:"-"`
+	Sender         string `json:"sender"`
+	Message        string `json:"message"`
+	CreatedDate    string `json:"created"`
+}
+
+type ConversationPost struct {
```
Comment (kzap) [inline internal/webserver/webserver.go]: specific struct for receiving POST data to `/messages`

File: internal/webserver/webserver.go [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+var routes []Route
+var db *sql.DB
+
+const webPort = 8000
+
+// Start starts the HTTP Server
+func Start(mysqlConfig config.DbConfig) {
+	var err error
+
+	routes = append(routes, Route{URL: "/", Description: "Homepage", Methods: "GET", Handler: GetIndex})
```
Comment (kzap) [inline internal/webserver/webserver.go]: Our URL Routes

File: internal/webserver/webserver.go [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+		return c, errors.New("JSON_EMPTY_MESSAGE")
+	}
+
+	return c, nil
+}
+
+// PostMessages handles the receiving of messages
+func PostMessages(w http.ResponseWriter, r *http.Request) {
+	contentType := r.Header.Get("Content-type")
+	switch contentType {
```
Comment (kzap) [inline internal/webserver/webserver.go]: accept different input data depending on the requested `Content-type`

File: internal/webserver/webserver.go [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+	routes = append(routes, Route{URL: "/messages", Description: "Message Input Form", Methods: "GET", Handler: GetMessages})
+	routes = append(routes, Route{URL: "/messages", Description: "Message POST Handler", Methods: "POST", Handler: PostMessages})
+
+	router := mux.NewRouter()
+	router.StrictSlash(true)
+
+	for _, route := range routes {
+		router.HandleFunc(route.URL, route.Handler).Methods(route.Methods)
+	}
+	router.PathPrefix("/").Handler(http.FileServer(http.Dir("./web/static"))).Methods("GET")
```
Comment (kzap) [inline internal/webserver/webserver.go]: load our static assets into the router

File: web/template/layout.html [hunk 9]
```diff
@@ -0,0 +1,47 @@
+{{define "_header"}}
+<!doctype html>
```
Comment (kzap) [inline web/template/layout.html]: Just used HTML5 Bootstrap and Bootstrap V4 CSS

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,38 @@
+version: '3.2'
+
+services:
+  web: &base
+    image: ada/backend-challenge:andre
+    build:
+      context: .
+      dockerfile: ./cmd/backend-challenge/Dockerfile
```
Comment (noqcks) [inline docker-compose.yml]: I think this wasn't checked in?

Comment (kzap) [reply docker-compose.yml]: Nice catch, sorry it was being .gitignore'd , I didnt test from a fresh clone

Everything is added now.

Comment (noqcks) [reply docker-compose.yml]: Np np, I'll review now.

File: deployments/docker-compose/docker-compose.dbdev.yml [hunk 11]
```diff
@@ -0,0 +1,21 @@
+# docker-compose -f docker-compose.yml -f deployments/docker-compose/docker-compose.dbdev.yml <command>
```
Comment (kzap) [inline deployments/docker-compose/docker-compose.dbdev.yml]: You can run this `docker-compose` command if you want to run phpMyAdmin and expose the MySQL port, useful for debugging in development

File: deployments/docker-compose/env/db.env [hunk 12]
```diff
@@ -0,0 +1,6 @@
+MYSQL_HOST=db
+MYSQL_HOST_PORT=db:3306
+MYSQL_DATABASE=ada_test
+MYSQL_USER=ada_test
+MYSQL_PASSWORD=ada_test
+MYSQL_ROOT_PASSWORD=ch@NgEth1s
```
Comment (kzap) [inline deployments/docker-compose/env/db.env]: the default MYSQL credentials

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+  db:
+    image: mysql:8.0
+    env_file:
+      - ./deployments/docker-compose/env/db.env
+    networks:
+      - ada_internal
+    volumes:
+      - db_mysql:/var/lib/mysql
+      - ./deployments/docker-compose/db/conversations.sql:/docker-entrypoint-initdb.d/conversations.sql
```
Comment (kzap) [inline docker-compose.yml]: this initializes the database on the first run

File: internal/cmd/root.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+	},
+}
+
+var MySQLConfig config.DbConfig
+
+func init() {
+	viper.SetEnvPrefix("ada")
+	viper.AutomaticEnv()
+	flags := rootCmd.PersistentFlags()
+	flags.StringVar(&MySQLConfig.Host, "db_host", "localhost", "mysql database host | env var: [ADA_DB_HOST]")
```
Comment (kzap) [inline internal/cmd/root.go]: our flags can also be set via ENV VARS

File: internal/webserver/webserver.go [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+	CreatedDate    string `json:"created"`
+}
+
+type ConversationPost struct {
+	ConversationID string `json:"conversation_id,omitempty"`
+	Sender         string `json:"sender"`
+	Message        string `json:"message"`
+}
+
+var templates = template.Must(template.ParseGlob(getCallerDir() + "/web/template/*.html"))
```
Comment (kzap) [inline internal/webserver/webserver.go]: preload our templates into memory

File: cmd/backend-challenge/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.12 as build-env
```
[the remark this answers is not in the document]
Comment (kzap) [reply cmd/backend-challenge/Dockerfile]: yep only issue is when debugging it, but [dive](https://github.com/wagoodman/dive) is useful for figuring out what exactly is inside the image

**Pull request closed without merging**